controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -509,17 +528,25 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
 }
 
 // Ensure we have the information about the hardware on the host.
-func (r *BareMetalHostReconciler) actionInspecting(prov provisioner.Provisioner, info *reconcileInfo) actionResult {
+func (r *BareMetalHostReconciler) actionInspecting(prov provisioner.Provisioner, info *reconcileInfo, forceNew bool) actionResult {
+	info.log.Info("inspecting hardware")
 
 	if inspectionDisabled(info.host) {
 		info.log.Info("inspection disabled by annotation")
 		info.publishEvent("InspectionSkipped", "disabled by annotation")
 		return actionComplete{}
 	}
 
-	info.log.Info("inspecting hardware")
+	// Delete inspect annotation if exists
+	if hasInspectAnnotation(info.host) {
+		if clearInspectAnnotation(info.host) {
+			if err := r.Update(context.TODO(), info.host); err != nil {
+				return actionError{errors.Wrap(err, "failed to remove inspect annotation from host")}
+			}
+		}
+	}
```

Comment:
+1 for removing `clearInspectAnnotation`.

Reply:
> The [original proposal](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md) states that the annotation will be removed by the controller once the inspection is completed. Not sure if I'm missing anything, but wouldn't be better to postpone the annotation cleanup _after_ a (succesfull) `InspectHardware` (also to retry in case of inspection error)?
> 
That's right. The reason I had to delete the annotation before calling [prov.InspectHardware ](https://github.com/metal3-io/baremetal-operator/blob/7b8f74a2f2e9d7c6d9fc08fd693d00997a229e73/controllers/metal3.io/baremetalhost_controller.go#L522) is because otherwise `forceNew` would always get `true`,  since the annotation is not deleted, so the host will be stuck in the re-inspection loop.

Reply:
I think that removing the `forceNew` will also simplify the code path, as the condition could be calculated directly here. In addition, removing the annotation in case of error could lead to unexpected results since the inspection will not be repeated

Reply:
Hi @andfasano . I have removed `forceNew` and tried to rely on existing `force`.
Calling `clearInspectAnnotation` just before
```
	clearError(info.host)
	info.host.Status.HardwareDetails = details
	return actionComplete{}
```
is causing endless inspection loop. Because the annotation isn't removed on time. Would you mind please to point out where would be the best to call it if not the current place where I added?

Reply:
I think you need to move it before line 628, otherwise you're exiting on line 633.

Reply:
> Hi @andfasano . I have removed `forceNew` and tried to rely on existing `force`.

Thanks @fmuyassarov but it looks like that collapsing the semantic of the two cases doesn't work properly. The "old" force flag was meant to address just the case where the node reached the `nodes.InspectFail` provisioning state, while the current case it's wider. Given that, it sounds probably a better idea to keep the two features separated for the sake of clarity (even though with a more complex `InspectHardware` signature). I'd like also to have a comment from @zaneb on this point

Reply:
Thank you @dtantsur , @andfasano . Updated the deletion.

Reply:
> Agree with @dtantsur , move it just before:
> 
> `if provResult.Dirty || details == nil {`

Thank you @dtantsur , @andfasano . Updated.